### PR TITLE
fix(#99): Envoy metered without CT reporting wrong values

### PR DESCRIPTION
**Pull request by catsmanac**
Fix for #99:

- Get watt_hours_lifetime from now_source in system_production.
- In api_v1_production check for 3 zero values for wattHour* instead of all values.
- In production remember all successful production pages and use first for production fallback, not last one used as this could be bad. 
- Add test fixture 7.3.466_metered_disabled_cts to test 
- Add the new test case to test_endpoints.

### CODE REVIEW

**Comment by catsmanac** [comment]
Did not include an updated snapshot as I ran into an issue I could not resolve. I'd appreciate some assistance here.

- The new code will detect that the /production and /production.json page have activeCount=0 for the eim.
- It will enable the fallback production to read wNow and whlifetime from the inverters section in the production page.
- A test program requesting data from simulated Envoy using flask returning the new test fixture 7.3.466_metered_disabled_cts data, shows /production in raw as well as the correct production numbers, with wNow and whLifetime from the inverters section.
- However, when running `poetry run pytest --snapshot-update` it seems to use the data from /production.json, rather then from /production.
  - The snapshots raw has the /production.json rather then the expected /production
  - The production values in the snapshot are those from /production.json
  - The timestamps in /production and /production.json differ slightly as well as some values, that triggered the new test to fail.

It's obvious I'm missing something, but so far no solution found.

Thanks for any help.

**Comment by catsmanac** [comment]
Moved the tests for the changed code out of the common V7 firmware test using snapshot and build them as topic tests. No snapshot changes with this approach and all tests now pass.

File: src/pyenphase/updaters/api_v1_production.py
Comment (bdraco) [inline src/pyenphase/updaters/api_v1_production.py:32]: This could use a comment explaining why `wattsNow` is not considered

File: src/pyenphase/models/system_production.py
Comment (bdraco) [inline src/pyenphase/models/system_production.py:50]: This could use a comment about why we use the `now_source` for `watt_hours_lifetime`

File: src/pyenphase/updaters/production.py
Comment (bdraco) [inline src/pyenphase/updaters/production.py:21]: If they have multiple systems this will likely cause a problem since the list isn't created in `__init__`

File: src/pyenphase/updaters/api_v1_production.py
Comment (catsmanac) [reply src/pyenphase/updaters/api_v1_production.py:32]: Added more comment in code.

File: src/pyenphase/models/system_production.py
Comment (catsmanac) [reply src/pyenphase/models/system_production.py:50]: It has the comment right above now_source calculation already.

```python
        # This is backwards compatible with envoy_reader
        # envoy metered without configured CT has whLifetime and wNow in inverters
        # whLastSevenDays and whToday are incorrect for both so either can be used
        now_source = eim if eim["activeCount"] else inverters
```

File: src/pyenphase/updaters/production.py
Comment (catsmanac) [reply src/pyenphase/updaters/production.py:21]: Thanks for cathing this one. Tested it, and indeed it showed up in the second device. Can't add it to `__init__` as it would reset it for the 2 subclasses as well and defeat the purpose. Changed code to reset the list in the fallback class when allow_inverters_fallback is true. That solved the issue.

And it also solved my problem with --snapshot-update saving /production.json rather then /production.

Comment (bdraco) [reply src/pyenphase/updaters/production.py:21]: If you have two unique systems setting up at the same time it's likely still a problem since one might clear the list out from under the other one

Comment (bdraco) [reply src/pyenphase/updaters/production.py:21]: I think you need to move this out of this class and into the main class to track it

Comment (catsmanac) [reply src/pyenphase/updaters/production.py:21]: My first thought would be to use common_properties pending in PR #108 to store this so it would be in the envoy data class and avoid the issue.

File: src/pyenphase/const.py
Comment (bdraco) [inline src/pyenphase/const.py:63]:
```suggestion
    PRODUCTION_FALLBACK_LIST = "production_fallback_list"
```

File: src/pyenphase/updaters/production.py
Comment (bdraco) [inline src/pyenphase/updaters/production.py:123]:
```suggestion
            CommonProperties.PRODUCTION_FALLBACK_LIST, self.working_endpoints
```

File: src/pyenphase/updaters/production.py
Comment (bdraco) [inline src/pyenphase/updaters/production.py:45]:
```suggestion
            self._common_properties.get(CommonProperties.PRODUCTION_FALLBACK_LIST) or []
```

File: src/pyenphase/updaters/production.py
Comment (bdraco) [inline src/pyenphase/updaters/production.py:26]: This still needs to be created in `__init__` to be safe so the list isn't shared

Comment (catsmanac) [reply src/pyenphase/updaters/production.py:26]: No need anymore as class variable. Made it local variable in probe method.

**Comment by bdraco** [comment]
https://github.com/pyenphase/pyenphase/pull/113 should fix the CI failure

I haven't had a chance to come back to do another review pass yet though

File: src/pyenphase/updaters/base.py
Comment (catsmanac) [reply src/pyenphase/updaters/base.py:28]: Ok, I'll add a common_properties.py in models defining a common_properties class

**Comment by bdraco** [comment]
LGTM.  I'll push it to much production systems later today and report back

**Comment by bdraco** [comment]
startup testing

Firmware: 3.9.36 ✅ 
Firmware: 7.3.605 ✅ 
Firmware: 5.0.62 ✅

**Comment by bdraco** [comment]
Sun just came up so should have data coming in within the next hour or so to compare

File: src/pyenphase/updaters/production.py
Comment (bdraco) [inline src/pyenphase/updaters/production.py:43]:
```suggestion
        # when allow_inverters_fallback mode is active use first successful endpoint registered in the list
```

**Comment by bdraco** [comment]
@catsmanac Thanks. Everything looks good

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
@catsmanac Since it seems like you plan on contributing longer term, I send you a github invite to `pyenphase`, if you choose to accept it, you should also add yourself to codeowners for the integration in HA.

**Comment by catsmanac** [comment]
> @catsmanac Since it seems like you plan on contributing longer term, I sent you a github invite to `pyenphase`, if you choose to accept it, you should also add yourself to codeowners for the integration in HA.

Thanks for your help and coaching. My plan is indeed to migrate some features (multiphase and grid IO) from a custom integration to core. If you feel joining makes matter simpler I'll join. I'll probably need more coaching in the future with only 9 months python under my belt. Not sure how this will change my activities. I assume I continue dev using my fork as before.

**Comment by bdraco** [comment]
> Thanks for your help and coaching. My plan is indeed to migrate some features (multiphase and grid IO) from a custom integration to core. If you feel joining makes matter simpler I'll join. 

Generally we prioritize reviews for members over someone first time contributors as its a bit easier when someone already has experience in the codebase.  All of us are volunteering our free time to work on this so time comes at a premium for most of us.

> I'll probably need more coaching in the future with only 9 months python under my belt. Not sure how this will change my activities. I assume I continue dev using my fork as before.

I wouldn't expect much to change in the workflow as we generally require a review before merging (except for small chores), except you could push branches to the repo here (if keep using your fork if you want), and you would be able to review and merge PRs from other members.

**Comment by catsmanac** [comment]
Thanks for your trust, I'll hope it can relive you more then it will cost.

Do I need to create a HA PR for the manifest to register as additional code-owner or is that something one of you will do?

**Comment by bdraco** [comment]
> Thanks for your trust, I'll hope it can relive you more then it will cost.
> 
> Do I need to create a HA PR for the manifest to register as additional code-owner or is that something one of you will do?

You'll need to open the PR to do that 

https://github.com/home-assistant/architecture/blob/master/adr/0008-code-owners.md

**Comment by bdraco** [comment]
@catsmanac if you send me a message and a friend request on discord (same handle), I can add you to our envoy chat with the other members.

**Pull request merged**
